src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializerHelpers.cs

```diff
@@ -0,0 +1,364 @@
+                    {
+                        deltasetItem.ODataIdContainer = container;
+                        deltasetItem.ODataPath = ODataPath;
+                    }
+                    else
+                    {
+                        PropertyInfo containerPropertyInfo = EdmLibHelpers.GetClrType(entityType, readContext.Model).GetProperties().Where(x => x.PropertyType == typeof(IODataIdContainer)).FirstOrDefault();
+                        if (containerPropertyInfo != null)
+                        {
+                            IODataIdContainer odataIdContainer = containerPropertyInfo.GetValue(resource) as IODataIdContainer;
+                            if (odataIdContainer != null)
+                            {
+                                containerPropertyInfo.SetValue(resource, odataIdContainer);
+                            }
+                            else
+                            {
+                                containerPropertyInfo.SetValue(resource, container);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// It parses a resource id, if it exists, as an OData Url and builds an ODataPath 
+        /// from the created URL.
+        /// </summary>
+        /// <param name="readContext">The read context.</param>
+        /// <param name="resourceWrapper">The resource wrapper.</param>
+        /// <returns>An ODataPath.</returns>
+        internal static Routing.ODataPath ApplyIdToPath(ODataDeserializerContext readContext, ODataResourceWrapper resourceWrapper)
+        {
+            // If an odata.id is provided, try to parse it as an OData Url.
+            // This could fail (as the id is not required to be a valid OData Url)
+            // in which case we fall back to building the path based on the current path and segments.
+            if (resourceWrapper.ResourceBase.Id != null)
+            {
+                IWebApiRequestMessage internalRequest = readContext.InternalRequest;
+                IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+                ODataPath odataPath = null;
+
+                if (internalRequest != null && urlHelper != null)
+                {
+                    string serviceRoot = urlHelper.CreateODataLink(
+                        internalRequest.Context.RouteName,
+                        internalRequest.PathHandler,
+                        new List<ODataPathSegment>());
+
+                    try
+                    {
+                        ODataUriParser parser = new ODataUriParser(readContext.Model, new Uri(serviceRoot), resourceWrapper.ResourceBase.Id);
+                        odataPath = parser.ParsePath();
+                    }
+                    catch (ODataException)
+                    {
+                        return null;
+                    }
+                    catch (UriFormatException)
```

Comment:
What scenario would give rise to this exception?

Reply:
passing an invalid URI when creating an instance of the URIPARSER - `new ODataUriParser(readContext.Model, new Uri(serviceRoot), resourceWrapper.ResourceBase.Id)`;